### DISCUSSION TITLE
Clean up physical layer

### DISCUSSION THREAD

**Discussion by aannleax** [knowsys/nemo#96] [Ideas]
The implementation of column scans is a bit inelegant. The trait `ColumnScan` defines the functions `pos` and `narrow` which only make sense for those column scans which directly iterate through a column. However, every other type of column scan still has to provide an implementation (currently just resorts to a `unimplemented!` call). Furthermore, every new type of `ColumnScan` has to be entered into the `ColumnScanEnum` object, which is tedious. A third issue is that in some cases, the type of the `ColumnScan` is known (e.g. in `TrieProject` one knows that the column scans are `ColumnScanReorder`), but they are still stored as a vector of `ColumnScanT`s. In order to limit the amount of "unpacking" one has to do, currently, all functions specific to one type of `ColumnScan` are also implemented on the `ColumnScanEnum` which panics if the function is called on the wrong type of column scan. However, it may be difficult to avoid this in general because sometimes we store several different types of `ColumnScan`s in the same vector, e.g. in `TrieJoin` the scans could either be `ColumnScanJoin` or `ColumnScanPass`.

To solving those problems you would probably:

- Instead of `ColumnScanEnum` use `dyn ColumnScan`
- Have `ColumnScanTypeT` enums for each type of column scan

Implementing the `dyn` approach can also be done first without the rest for testing the performance impact. In this case one would leave the extra functions `pos` and `narrow` in the trait.

**Comment by matzemathics** [comment]
This crate might be of interest for cleaning up the glue code without compromising on performance: https://docs.rs/enum_dispatch/latest/enum_dispatch/
